GCSClient

benjamincerigo opened the pull request:
This is a pre PR for a PR is made on the original repo.

## Includes:
- Addition of the GCS client
- Implementation of the `get` and `put` methods plus unit test for these. 
- `project` and `creds` arguments
- a functional test that will write to a bucket as defined in the `tests/functional/gcs/conftest.py`.

## Some questions:
should it be `gcs` or `gs`:
- `gs` usage example: [by google](https://cloud.google.com/storage/docs/gsutil)
- `gcs` usage: [kedro](https://kedro.readthedocs.io/en/stable/04_user_guide/08_advanced_io.html?highlight=gcs#checkpoint-configuration), [fsspec](https://github.com/intake/filesystem_spec#purpose)

benjamincerigo committed:
GCS: google-cloud-storage install

benjamincerigo committed:
GCS: GSClient init

Initialisation of the GSClient implemented and unit tested.

benjamincerigo committed:
GCS: get implemented

benjamincerigo committed:
GSC: refactor into _get_blob

benjamincerigo committed:
GCS: put method implemented

benjamincerigo committed:
GCS: project and creds args

Addition of the project and creds args to the GCSClient.
These will be passed to the storage client.
If non are set then the default are gotten from the environment.

benjamincerigo committed:
GCS: functional testing

Functional test to the test api

DBCerigo commented on setup.py:
```diff
@@ -34,6 +34,9 @@ def run(self):
     # AWS
     "boto3>=1.9.0,<1.10",
+    # GCS
+    "google-cloud-storage<1.29",
+    "google-api-core<1.18",
```
How did you decide to constrict the versions like this? 

I think if you don't have a good reason to (i.e. I know api breaking changes occurred at this version), then youo shouldn't constrain the versions.

DBCerigo commented on Makefile:
```diff
@@ -36,6 +36,11 @@ unit:
 	pipenv run pytest tests/functional/s3
 
+functional-gcs:
+	# You will need to have the authentication set up to run these test
+	# 🚨🚨🚨WARNING🚨🚨🚨 -  This will create a bucket and file as configured in tests/functional/gcs/conftest.py
```
Does these same considerations not also apply for the `s3` tests also? If so, then don't add these comments, as there aren't comments like these for the `s3` ones.

ali-tny replied:
The s3 functional tests use [`moto`, an in-memory S3 emulator,](https://github.com/spulec/moto) so no buckets/files are created. Unfortunately no such library exists for google cloud, and according to various issues ([eg](https://github.com/googleapis/google-cloud-python/issues/5816)) the google cloud python team don't see it as in their remit, so its unlikely to be around any time soon. It's kind of a bummer to have to actually create files but I guess it's a necessary evil - Javi might have some input if there's a better way.

benjamincerigo replied:
I am also not sure about this and depends on what the maintainers want in there tests.

DBCerigo reviewed (commented):
Very very nice 👌


> ## Some questions:
> should it be `gcs` or `gs`:
> 
> * `gs` usage example: [by google](https://cloud.google.com/storage/docs/gsutil)
> * `gcs` usage: [kedro](https://kedro.readthedocs.io/en/stable/04_user_guide/08_advanced_io.html?highlight=gcs#checkpoint-configuration), [fsspec](https://github.com/intake/filesystem_spec#purpose)

Even though there are those two examples, I think it must be `gs`, anything else is a bizarre departing from the actual service's API it relies on. (Yes I think that statement applies to `kedro` and `fsspec`. Similarly the class and var names etc, should all have the `C` or `c` removed.

ali-tny commented on src/tentaclio/clients/gcs_client.py:
```diff
@@ -0,0 +1,160 @@
+    Ref: https://googleapis.dev/python/storage/latest/index.html
+    """
+
+    allowed_schemes = ["gcs"]
```
re the URI scheme, I think it makes sense to use the official APIs scheme `gs`, and via a quick google search seems like thats the only one I immediately see, but if there's some other often-used scheme then we could just include both here (see [the LocalFileClient](https://github.com/datavaluepeople/tentaclio/blob/master/src/tentaclio/clients/local_fs_client.py) for precedent - both schemes are also registered to the client in the `STREAM_HANDLER_REGISTRY`)

benjamincerigo replied:
Done

ali-tny commented on src/tentaclio/clients/gcs_client.py:
```diff
@@ -0,0 +1,160 @@
+        """
+        gcs_bucket, gcs_key = self._fetch_bucket_and_key(bucket_name, key_name)
+
+        try:
+            self._put(reader, gcs_bucket, gcs_key)
+        except google_exceptions.NotFound:
+            raise exceptions.GCSError("Unable to fetch the remote file")
```
I think an essential workflow is to write to a new file, so the put method should support that

benjamincerigo replied:
Incorrect message in exception. Is there to catch the bucket not found.

ali-tny commented on tests/functional/gcs/conftest.py:
```diff
@@ -0,0 +1,58 @@
+
+from tentaclio import clients
+
+TEST_BUCKET = "tentaclio-bucket"
+
+GCS_TEST_URL = os.getenv("TENTACLIO__CONN__GCS_TEST")
+GCS_TEST_PROJECT = os.getenv("TENTACLIO__CONN__GCS_TEST_PROJECT")
```
Should we infer the bucket name from the test URL? otherwise they have to be in sync here right?

benjamincerigo replied:
Good one. Done

ali-tny commented on tests/unit/clients/test_gcs_client.py:
```diff
@@ -0,0 +1,230 @@
+    gcs_client.creds = "set_creds"
+
+
+@mock.patch("tentaclio.clients.GCSClient._get")
+@mock.patch("tentaclio.clients.GCSClient._init_creds")
+@mock.patch("tentaclio.clients.GCSClient._connect")
+@pytest.mark.parametrize(
+    "url,bucket,key", [
+        ("gcs://:@gcs", None, None),
+        ("gcs://:@gcs", "bucket", None),
+        ("gcs://:@bucket", None, None)
+    ],
+)
+def test_get_invalid_path(m_connect, m_init, m_get, url, bucket, key):
+    """Test get for with invalid paths."""
+    with gcs_client.GCSClient(url) as client:
+        with pytest.raises(exceptions.GCSError):
+            client.get(io.StringIO(), bucket_name=bucket, key_name=key)
+
+    m_get.assert_not_called()
```
most of the put/get tests are the same other than the method being called - might be a bit less verbose to parametrize the method, what ya think?

benjamincerigo replied:
Done. I thought about doing this when I was reviewing it but I felt that it might be harder to understand. Seems to be much more understandable.

ali-tny commented on src/tentaclio/clients/gcs_client.py:
```diff
@@ -0,0 +1,160 @@
+        project: Optional[str] = None,
+        creds: Optional[credentials.Credentials] = None
+    ) -> None:
+        default_creds, default_project = self._get_default_creds()
+        self.creds = creds or default_creds
+        self.project = project or default_project
+
+        if self.project is None:
+            raise exceptions.GCSError("Must have a project set to use gcs.")
+
+        if self.creds is None:
+            raise exceptions.GCSError("Must have credentials to use gcs.")
```
I think that `google.auth.default` raises an exception if there are no default creds, so 
a) if the creds are set manually & there are no defaults it will (wrongly) fail
b) the exceptions.GCSError will never get thrown - instead the google client library will throw the google error

In fact, I think that `storage.Client` will call `default` under the hood if the credentials / project is left to default anyway, so you avoid doing the work here in `_init_creds` and leave it all to figure it out in `_connect`. Could leave the `google.auth.DefaultCredentialsError` there, or wrap it in a tentaclio exception.

ali-tny commented on src/tentaclio/clients/gcs_client.py:
```diff
@@ -0,0 +1,160 @@
+        self,
+        url: Union[urls.URL, str],
+        project: Optional[str] = None,
+        creds: Optional[credentials.Credentials] = None
```
Passing the credentials like this isn't really canonical tentaclio, because we'd prefer to have the primary method of authentication to be embedded in the URL. This is what we do in S3 as well - but in practice, no one stores their S3 credentials in tentaclio but rather they've already logged in to the aws cli, which stores the credentials in a default place that boto (the s3 client library) knows how to read.

So, this will in practice do the same thing: the google storage client will read the credentials from the default place (and in production, the file will be placed or an environment variable set or whatever). However it'd be nice to be able to authenticate the "usual" way as well. I had a look and google uses OAuth2, unfortunately requiring an id, a secret _and_ a refresh token, and I'm not sure how that would fit into the URI authority format.

Default behaviour I think is to consider anything after the `:` and before the `@` as secret in `scheme:user:pass@host:port/path`, so we could shoehorn both tokens in there with a separator or something... but maybe thats too much of a hack. Javi might have some thoughts or he might just think its chillin to leave the auth to the google client library. Maybe when you make the PR to the main repo, add a line in the description about the google auth situation

benjamincerigo replied:
[See comment](https://github.com/datavaluepeople/tentaclio/pull/1#issuecomment-635339555)

ali-tny reviewed (changes requested):
looks great! some comments about changes (particularly the `put` method) and some other comments that are more just discussion

benjamincerigo committed:
fixup! GCS: put method implemented

Refactor the get and put tests so that they use parameters

benjamincerigo committed:
fixup! GCS: get implemented

More asserts on the _get test. So that it will be different from the
_put test (see next commit).

benjamincerigo committed:
fixup! GCS: put method implemented

Asserts added to the _put test.

benjamincerigo committed:
fixup! GCS: google-cloud-storage install

Remove of the version for google-cloud-storage.

benjamincerigo committed:
fixup! GCS: GSClient init

"gcs" as allowed schema for GSClient

benjamincerigo committed:
fixup! GCS: put method implemented

Error message fix for bucket not found

benjamincerigo committed:
fixup! GCS: functional testing

Test bucket taken from the test URL.

benjamincerigo committed:
fixup! GCS: functional testing

Remove of the unneed project from the functional tests

benjamincerigo committed:
Revert "GCS: project and creds args"

This reverts commit f811f188ba3fd2b53b936fb27f757f1957b3ca14.

benjamincerigo committed:
fixup! GCS: GSClient init

benjamincerigo commented:
Have now pushed based on the reviews. 

## gcs -> gs
I have done a string replace of GCS and gcs to GS and gs. I deiced to just rewrite the original commits to do this since it would have been a complex to add these as new commit.

If you just want to start checking the commits from `ad68a84` then you can see all the other changes as fix up commits. 

## project and creds args
I have reverted the `GCS: project and creds args` commit. Originally I was having a problem with the setting of the project for the storage client. However, after reading the docs again I realised that this should also be set with an env variable. As such, I added this to the docs of the init string with a fixup commit. 

My feeling on this is that since googles docs seem to only use env variables the Client does not need to implement any passwords or usernames in the URL (or args of init). This would then be in keeping with the docs.

## Execptions
After some consideration, I felt that it was best to not catch the errors from the storage client. This way the user of the API gets more descriptive exceptions and we don't have to repeat the same exceptions.  

## Remove function was also added.

benjamincerigo committed:
fixup! GCS: functional testing

benjamincerigo committed:
GCS: remove added

benjamincerigo committed:
fixup! GCS: get implemented

Remove of the expcetion catch for the get

benjamincerigo committed:
fixup! GCS: get implemented

Improved linting of the fetch function

benjamincerigo committed:
fixup! GCS: put method implemented

Remove of the catched expection in the put

benjamincerigo committed:
GCS: remove method added

benjamincerigo committed:
GCS: valid method test with URL only

ali-tny commented on src/tentaclio/clients/gs_client.py:
```diff
@@ -0,0 +1,156 @@
+    def _remove(
+        self, bucket_name: str, key_name: str
+    ) -> None:
+        """Upload on the client."""
```
pasted docstring

benjamincerigo replied:
Done

ali-tny commented on tests/unit/clients/test_gs_client.py:
```diff
@@ -0,0 +1,208 @@
+
+
+@mock.patch("tentaclio.clients.GSClient._connect")
+@pytest.mark.parametrize(
+    "method,url,bucket,key", [
+        ("get", "gs://:@gs", None, None),
+        ("get", "gs://:@gs", "bucket", None),
+        ("get", "gs://:@bucket", None, None),
+        ("put", "gs://:@gs", None, None),
+        ("put", "gs://:@gs", "bucket", None),
+        ("put", "gs://:@bucket", None, None),
+        ("remove", "gs://:@gs", None, None),
+        ("remove", "gs://:@gs", "bucket", None),
+        ("remove", "gs://:@bucket", None, None)
+    ],
+)
```
fyi can do cartesian products of parameters:
```
@pytest.mark.parametrize("method", ["get", "put", "remove"])
@pytest.mark.parametrize(
    "url,bucket,key",
    [("gs//:@gs", None, None), ("gs://:@gs", "bucket", None), ("gs://:@bucket", None, None)],
)
def test_invalid_path(m_connect, method, url, bucket, key):
    ...
```
should be able to remove boilerplate on most of these tests functions this way

benjamincerigo replied:
🤯 .

ali-tny commented on tests/unit/clients/test_gs_client.py:
```diff
@@ -0,0 +1,208 @@
+    assert client.bucket == hostname
+
+
+def patch_string_from_method(method: str) -> str:
+    """Get the string that indicates the method to be patched for a calling method."""
+    switch_dict = {
+        "get": "tentaclio.clients.GSClient._get",
+        "put": "tentaclio.clients.GSClient._put",
+        "remove": "tentaclio.clients.GSClient._remove"
+    }
+
+    if method in switch_dict:
+        return switch_dict[method]
+
+    raise Exception(f"No patch string for calling method: {method}")
```
could avoid this function entirely and just do `mocker.patch(f"tentaclio.clients.GSClient._{method}")` in the tests and it'll raise a mocker exception if it's missing, provided you use the pytest `mocker` fixture instead of `mock` - just need to pass an argument `mocker` to the test functions that use mock and it'll do the rest

ali-tny commented on tests/unit/clients/test_gs_client.py:
```diff
@@ -0,0 +1,208 @@
+    """Test a method with invalid paths."""
+    patch_string = patch_string_from_method(method)
+
+    with mock.patch(patch_string) as mocked_method:
+        with gs_client.GSClient(url) as client:
+            with pytest.raises(exceptions.GSError):
```
FYI you can combine context managers into one line so you don't have to do 3 indents, and the mock can be safely done without context manager because it's only used as mock throughout the test:
```
mocked_method = mocker.patch(patch_string)
with gs_client.GSClient(url) as client, pytest.raises(exceptions.GSError):
    ...
```

ali-tny commented on tests/functional/gs/test_api.py:
```diff
@@ -0,0 +1,25 @@
+from google.cloud import exceptions as google_exceptions
+
+
+def test_authenticated_api_calls(gs_url, fixture_client):
+    """Test the authetnciated api calls."""
+    data = bytes("Ph'nglui mglw'nafh Cthulhu R'lyeh wgah'nagl fhtagn", "utf-8")
+
+    with tentaclio.open(gs_url, mode="wb") as f:
+        f.write(data)
+
+    with tentaclio.open(gs_url, mode="rb") as f:
+        result = f.read()
+
+    assert result == data
+
+    tentaclio.remove(gs_url)
+
+    with pytest.raises(google_exceptions.NotFound):
+        with tentaclio.open(gs_url, mode="rb") as f:
+            result = f.read()
```
nice 👍

ali-tny commented on src/tentaclio/clients/gs_client.py:
```diff
@@ -0,0 +1,156 @@
+            GSError: If a bucket or a key is not found in the given URL or args then
+            Google Cloud Exceptions: if the client raises them.
+        """
+        gs_bucket, gs_key = self._fetch_bucket_and_key(bucket_name, key_name)
+
+        self._get(writer, gs_bucket, gs_key)
```
looking at these get/put/remove functions, you could probably just combine the hidden methods with these and then the all the tests that mock the private methods can just be tests of _fetch_bucket_and_key. i guess you get a tiny bit lower coverage wrt passing the bucket + key along to the next part of the function, and maybe its harder to test the private methods tho... so i dont really mind either way

ali-tny reviewed (approved):
lookin good, just some comments primarily that could clean up the testing some more

DBCerigo reviewed (approved):
All sweet with me

benjamincerigo committed:
fixup! GCS: remove added

_remove method docstring fix

benjamincerigo committed:
fixup! GCS: put method implemented

gs_test clean up of params

benjamincerigo committed:
GCS: with combine

Needs to be fixup by hand

benjamincerigo commented:
Awesome! I am going to close this PR and will open one up in the octoengy repo.
